patches/server/0854-fixup-Highly-optimise-single-and-multi-AABB-VoxelSha.patch

```diff
@@ -0,0 +1,720 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Spottedleaf <Spottedleaf@users.noreply.github.com>
+Date: Fri, 14 Jan 2022 13:14:31 -0800
+Subject: [PATCH] fixup! Highly optimise single and multi-AABB VoxelShapes and
+ collisions
+
+
+diff --git a/src/main/java/io/papermc/paper/util/CollisionUtil.java b/src/main/java/io/papermc/paper/util/CollisionUtil.java
+index 58629451977c89db2fa895bde946135784a0d8bc..3ce132ea28ff9fc6f7afd68bcd0fc1ae843a3cd5 100644
+--- a/src/main/java/io/papermc/paper/util/CollisionUtil.java
++++ b/src/main/java/io/papermc/paper/util/CollisionUtil.java
+@@ -15,6 +15,7 @@ import net.minecraft.world.level.block.state.BlockState;
+ import net.minecraft.world.level.border.WorldBorder;
+ import net.minecraft.world.level.chunk.ChunkAccess;
+ import net.minecraft.world.level.chunk.LevelChunkSection;
++import net.minecraft.world.level.chunk.PalettedContainer;
+ import net.minecraft.world.level.material.FlowingFluid;
+ import net.minecraft.world.level.material.FluidState;
+ import net.minecraft.world.phys.AABB;
+@@ -25,7 +26,6 @@ import net.minecraft.world.phys.shapes.EntityCollisionContext;
+ import net.minecraft.world.phys.shapes.Shapes;
+ import net.minecraft.world.phys.shapes.VoxelShape;
+ import java.util.List;
+-import java.util.Optional;
+ import java.util.function.BiPredicate;
+ import java.util.function.Predicate;
+ 
+@@ -33,6 +33,30 @@ public final class CollisionUtil {
+ 
+     public static final double COLLISION_EPSILON = 1.0E-7;
+ 
++    public static final long KNOWN_EMPTY_BLOCK = 0b00; // known to always have voxelshape of empty
++    public static final long KNOWN_FULL_BLOCK = 0b01; // known to always have voxelshape of full cube
++    public static final long KNOWN_UNKNOWN_BLOCK = 0b10; // must read the actual block state for info
++    public static final long KNOWN_SPECIAL_BLOCK = 0b11; // caller must check this block for special collisions
++
++    public static final List<String> BAD_PLUGINS = java.util.Arrays.asList(
++            "FastAsyncWorldEdit"
```

Comment:
Why is this plugin specifically singled out here?

Reply:
Because it does stupid things to chunks.

Reply:
To add an actual explanation: FAWE does not use the very inefficient methods to set single block states in chunk sections. Instead, ticking blocks are updated all at once per chunk section. Similar to every plugin that e.g. uses the approach 4 describe in https://www.spigotmc.org/threads/methods-for-changing-massive-amount-of-blocks-up-to-14m-blocks-s.395868/, this will not call the methods required by this patch.

Reply:
to my understanding this concern was actually resolved so no longer an issue here